fix(torch): to accommodate CUDA-13 and nvidia_cuda_runtime-13.0.48 wh…

whitesscott opened the pull request:
…eel preload CUDA deps from wheels

Replace /proc/self/maps check with an importlib.metadata scan to detect libcudart.so* shipped by wheels. CUDA-13 directory tree nvidia/cu13/lib and earlier CUDA slim-wheel paths. Preload nvrtc and nvJitLink accordingly. Preserve Linux-only behavior..

Rationale:
- /proc/self/maps depends on runtime load state and misses wheel-provided libs.
- importlib.metadata is environment-agnostic (venv/system/uv/conda) and robust to layout changes.

Risk:
- Low. Only affects preloading path for wheel-based CUDA installs; no change on Windows/macOS.

whitesscott committed:
fix(torch): to accommodate CUDA-13 and nvidia_cuda_runtime-13.0.48 wheel preload CUDA deps from wheels

Replace /proc/self/maps check with an importlib.metadata scan to detect libcudart.so* shipped by wheels.
CUDA-13 directory tree nvidia/cu13/lib and includes CUDA-12 slim-wheel paths.
Preload nvrtc and nvJitLink accordingly. Preserve Linux-only behavior..

Rationale:
- /proc/self/maps depends on runtime load state and misses wheel-provided libs.
- importlib.metadata is environment-agnostic (venv/system/uv/conda) and robust to layout changes.

Risk:
- Low. Only affects preloading path for wheel-based CUDA installs; no change on Windows/macOS.

Signed-off-by: scott white <whitesscott@comcast.net>

whitesscott commented:
@pytorchbot label "topic: not user facing"

whitesscott committed:
Fix indentation and exception ordering in _load_global_deps

Skylion007 commented on torch/__init__.py:
```diff
@@ -368,10 +378,16 @@ def _load_global_deps() -> None:
+
         ctypes.CDLL(global_deps_lib_path, mode=ctypes.RTLD_GLOBAL)
 
+    except Exception:
```
Ugh... this is dangerous. Can we at least log the exception or something?! Shouldn't this be SystemError, ImportError or something more specific?

Skylion007 replied:
I mean shouldn't we want to hard file sometimes?

whitesscott committed:
fix(torch): make torch/__init__.py mypy-clean and lintrunner-clean

Use builtins.any to avoid torch.any overload resolution and annotate with builtins.bool to avoid torch._C.bool in type positions. Also minor typing cleanup around distribution file iteration.

Verification: ran locally — mypy -m torch --check-untyped-defs and lintrunner --all-files both passed. These checks will continue to be used going forward.

whitesscott committed:
Merge branch 'pytorch:main' into fix/torch-cuda13-linux
